assets/js/modules/analytics-4/utils/withCustomDimensions.js

```diff
@@ -0,0 +1,374 @@
+import Link from '../../../components/Link';
+import MetricTileError from '../../../components/KeyMetrics/MetricTileError';
+import MetricTileWrapper from '../../../components/KeyMetrics/MetricTileWrapper';
+import {
+	ERROR_CODE_MISSING_REQUIRED_SCOPE,
+	ERROR_REASON_BAD_REQUEST,
+	isInsufficientPermissionsError,
+} from '../../../util/errors';
+const { useSelect, useDispatch } = Data;
+
+export default function withCustomDimensions( options = {} ) {
+	const { dimensions, infoTooltip, reportOptions, title } = options;
+
+	return ( WrappedComponent ) => {
+		const WithCustomDimensionsComponent = ( props ) => {
+			const { Widget, widgetSlug } = props;
+			const {
+				description,
+				infoTooltip: definedInfoTooltip,
+				requiredCustomDimensions,
+				title: definedTitle,
+			} = KEY_METRICS_WIDGETS[ widgetSlug ] || {};
+
+			const tileTitle = title || definedTitle;
+			const tileInfoTooltip =
+				infoTooltip || definedInfoTooltip || description;
+
+			const customDimensions = useMemo( () => {
+				if ( Array.isArray( dimensions ) && dimensions.length ) {
+					return dimensions;
+				}
+
+				if (
+					Array.isArray( requiredCustomDimensions ) &&
+					requiredCustomDimensions.length
+				) {
+					return requiredCustomDimensions;
+				}
+
+				return null;
+			}, [ requiredCustomDimensions ] );
+
+			const hasCustomDimensions = useSelect(
+				( select ) =>
+					! customDimensions ||
+					select( MODULES_ANALYTICS_4 ).hasCustomDimensions(
+						customDimensions
+					)
+			);
+			const isCreatingCustomDimensions = useSelect(
+				( select ) =>
+					!! customDimensions &&
+					customDimensions.some( ( dimension ) =>
+						select( MODULES_ANALYTICS_4 ).isCreatingCustomDimension(
+							dimension
+						)
+					)
+			);
+			const customDimensionsCreationErrors = useSelect( ( select ) => {
+				if ( ! customDimensions ) {
+					return [];
+				}
+
+				const errors = [];
+
+				customDimensions.forEach( ( dimension ) => {
+					const error =
+						select(
+							MODULES_ANALYTICS_4
+						).getCreateCustomDimensionError( dimension );
+
+					if ( error ) {
+						errors.push( error );
+					}
+				} );
+
+				return errors;
+			} );
+			const helpLink = useSelect( ( select ) => {
+				if ( ! customDimensionsCreationErrors.length ) {
+					return undefined;
+				}
+
+				return select( CORE_SITE ).getErrorTroubleshootingLinkURL(
+					customDimensionsCreationErrors.find(
+						isInsufficientPermissionsError
+					) || customDimensionsCreationErrors[ 0 ]
+				);
+			} );
+			const hasAnalyticsEditScope = useSelect( ( select ) => {
+				if ( ! customDimensions ) {
+					return undefined;
+				}
+
+				return select( CORE_USER ).hasScope( ANALYTICS_EDIT_SCOPE );
+			} );
+			const isSyncingAvailableCustomDimensions = useSelect(
+				( select ) => {
+					if ( ! customDimensions ) {
+						return false;
+					}
+
+					return select(
+						MODULES_ANALYTICS_4
+					).isSyncingAvailableCustomDimensions();
+				}
+			);
```

Comment:
We can use the same shorthands here to keep this component a bit more concise:

```suggestion
			const hasAnalyticsEditScope = useSelect( ( select ) =>
				customDimensions &&
				select( CORE_USER ).hasScope( ANALYTICS_EDIT_SCOPE )
			);
			const isSyncingAvailableCustomDimensions = useSelect(
				( select ) =>
					customDimensions &&
					select( MODULES_ANALYTICS_4 ).isSyncingAvailableCustomDimensions()
			);
```

Reply:
The same applies to all selectors in this file. This will let us have less lines in this component.

Reply:
Updated, thanks!